src/shogun/machine/LinearMachine.h

```diff
@@ -142,6 +134,18 @@ class LinearMachine : public Machine
 		 */
 		virtual SGVector<float64_t> apply_get_outputs(std::shared_ptr<Features> data);
 
+		bool train_machine(const std::shared_ptr<Features>& data, const std::shared_ptr<Labels>& labs) final
+		{
+			const auto dot_feat = std::dynamic_pointer_cast<DotFeatures>(data);
+			require(dot_feat, "Expected DotFeatures type");
```

Comment:
I think here you either use a more descriptive error message, or you rewrite this as `const auto dot_feat = data->as<DotFeatures>()`